compiler/src/dotty/tools/dotc/transform/TupleRewrites.scala

```diff
@@ -0,0 +1,215 @@
+            sndPat match {
+              case Literal(Constant(())) =>
+                  Some((List(fstPat), UnfoldedTupleType(fstTpe.tpe :: Nil)))
+
+              case UnApply(TypeApply(Select(ident, nme.unapply), tailTpes), Nil, tailPats)
+                if defn.DottyTupleNModule contains ident.symbol =>
+                  Some((fstPat :: tailPats, UnfoldedTupleType(fstTpe.tpe :: tailTpes.map(_.tpe))))
+
+              case UnApply(TypeApply(Select(ident, nme.unapplySeq), tailTpes), Nil, tailPat)
+                if ident.symbol == defn.TupleUnapplySeqModule  =>
+                  val foldedTailType = defn.TupleConsType.safeAppliedTo(tailTpes.map(_.tpe))
+                  Some((fstPat :: tailPat, FoldedTupleType(fstTpe.tpe, foldedTailType)))
+
+              case Typed(UnApply(TypeApply(Select(ident, nme.unapply), tailTpes), Nil, tailPats), _)
+                if defn.DottyTupleNModule contains ident.symbol =>
+                  Some((fstPat :: tailPats, UnfoldedTupleType(fstTpe.tpe :: tailTpes.map(_.tpe))))
+
+              case Typed(UnApply(TypeApply(Select(ident, nme.unapplySeq), tailTpes), Nil, tailPats), _)
+                if ident.symbol == defn.TupleUnapplySeqModule  =>
+                  val foldedTailType = defn.TupleConsType.safeAppliedTo(tailTpes.map(_.tpe))
+                  Some((fstPat :: tailPats, FoldedTupleType(fstTpe.tpe, foldedTailType)))
+
+              case _ => None
+            }
+        case _ => None
+      }
+  }
+
+  /** Rewrites `TupleCons.unapply(a, TupleCons.unapply(b, ..., TNit))` to implementation specific extractors.
+   *
+   *  Below `MaxImplementedTupleArity`, they become `DottyTuple$i.unapply(a, b, ...)`.
+   *  Above `MaxImplementedTupleArity`, they become `TupleUnapplySeq.unapply(a, b, ...)`.
+   *
+   *  Similarly to `transformApply`, size `N` extractors will pass `N` times thought this transformation.
+   */
+  override def transformUnApply(tree: UnApply)(implicit ctx: Context, info: TransformerInfo): Tree =
+    tree match {
+      case TupleUnapplies(patterns, types) =>
+        transformUnApplyPatterns(tree, patterns, types)
+      case _ => tree
+    }
+
+  /** Same then `transformUnApply` for unapply wrapped in Typed trees. */
+  override def transformTyped(tree: Typed)(implicit ctx: Context, info: TransformerInfo): Tree =
+    tree match {
+      case Typed(TupleUnapplies(patterns, types), tpe) =>
+        val unapply = transformUnApplyPatterns(tree, patterns, types)
+        Typed(unapply, tpe)
+      case _ => tree
+    }
+
+  // Create an `UnApply` tree from a list of patters, used in both transformUnApply and transformTyped.
+  private def transformUnApplyPatterns(tree: Tree, patterns: List[Tree], types: TupleType)(implicit ctx: Context): UnApply = {
+    val arity = patterns.length
+    if (arity <= MaxImplementedTupleArity) {
+      val unfoldedTypes: List[Type] = types.unfolded.types
+      val refinedType  = defn.TupleNType(arity).safeAppliedTo(unfoldedTypes)
```

Comment:
`appliedTo`, unless you encountered a `CyclicReference` -- but I would be surprised if you did! Note that `safeAppliedTo` is not uniformly better than `appliedTo` -- in fact it is worse because it makes up hypothetical types instead of forcing.